### PR TITLE
clh: Add TDX support

### PR DESCRIPTION
**Pull request by fidencio**
TDX support using td-shim has been added to cloud-hypervisor not so long ago, and in this series I'm bringing that functionality to the cloud-hypervisor driver.

Fixes: #3632

### CODE REVIEW

**Comment by fidencio** [comment]
I am adding the `wip` label till I run the last round of tests and collect the info on what's the components reviewers can use to test it.

/test

**Comment by fidencio** [comment]
CI has gone awry about the cloud-hypervisor builds as there are changes we bring in as part of the cloud-hypervisor bump that affect how we build the project.

**Comment by fidencio** [comment]
https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3773 was proposed to Cloud Hypervisor.

**Comment by fidencio** [comment]
/test

So we can at least ensure there's no breakage on the CIs not using Cloud Hyperisor.

**Comment by fidencio** [comment]
As https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3773 got fixed, I'm updating this PR to bring that fix in.

**Review by devimc — APPROVED** [review]
thanks @fidencio :rocket:

**Review by likebreath — APPROVED** [review]
LGTM. I wonder is it the plan that we will solely use and ship Kata with tdx-enabled Cloud Hypervisor binary? 

If that's the plan, can you please update the test script to build the tdx-enabled binary and make sure the CI works? Thank you.

**Comment by fidencio** [comment]
> LGTM. I wonder is it the plan that we will solely use and ship Kata with tdx-enabled Cloud Hypervisor binary?

Yep, as that doesn't have any negative impact on non confidential guests.

> If that's the plan, can you please update the test script to build the tdx-enabled binary and make sure the CI works? Thank you.

We already do that. :-)
https://github.com/kata-containers/kata-containers/pull/3771/files#diff-4e9dffeafb07dc536a419fa1c12eda2ed1fba26d98f3987694a13fafecc4362dR130-R132 is the part where we force `--features tdx` to be used.  And this script is used as part of kata-deploy, our GitHub actions, and also our CI.  So, we should be covered on all fronts there.

**Review by sboeuf — APPROVED** [review]
Looks good!

**Comment by likebreath** [comment]
> We already do that. :-) https://github.com/kata-containers/kata-containers/pull/3771/files#diff-4e9dffeafb07dc536a419fa1c12eda2ed1fba26d98f3987694a13fafecc4362dR130-R132 is the part where we force `--features tdx` to be used. 

oh, right. I missed it. You may want to split the scripts changes from the rest of clh version bump (the first commit).

> And this script is used as part of kata-deploy, our GitHub actions, and also our CI. So, we should be covered on all fronts there.

This is great.

**Comment by fidencio** [comment]
> oh, right. I missed it. You may want to split the scripts changes from the rest of clh version bump (the first commit).

I thought about doing that, but the bump broke the current scripts we currently have, thus I decided to the changes that are needed due to the changes in the script as part of the bump.

I am super fine about splitting as well, if you think it'd be better.  Please, just let me know your preference, @likebreath.

**Comment by likebreath** [comment]
> I thought about doing that, but the bump broke the current scripts we currently have, thus I decided to the changes that are needed due to the changes in the script as part of the bump.

I see. I agree having non-breaking commit is more preferable. Let's keep it in this way. Thanks for the context.

Looks like the CIs are all going green. Anything else you are holding before we can land it? @fidencio

**Comment by fidencio** [comment]
> Looks like the CIs are all going green. Anything else you are holding before we can land it? @fidencio

There are some last tests I'd like to run (locally) before getting this one merged.

OTOH, at this point I think we can go ahead with this one in order to avoid blocking the "-rc0" and then fix whatever has to be fixed (and backport the patches) after "-rc0", in case we find out about some issue.

File: src/runtime/virtcontainers/clh.go [hunk 1]
```diff
@@ -606,6 +656,10 @@ func (clh *cloudHypervisor) HotplugRemoveDevice(ctx context.Context, devInfo int
 	span, _ := katatrace.Trace(ctx, clh.Logger(), "HotplugRemoveDevice", clhTracingTags, map[string]string{"sandbox_id": clh.id})
 	defer span.End()
 
+	if clh.config.ConfidentialGuest {
+		return nil, errors.New("Device hotplug addition is not supported in confidential mode")
```
Comment (egernst) [inline src/runtime/virtcontainers/clh.go:660]: nit -- s/addition/removal

File: src/runtime/virtcontainers/hypervisor.go [hunk 2]
```diff
@@ -564,6 +564,11 @@ func (conf *HypervisorConfig) Valid() error {
 		conf.DefaultMaxVCPUs = defaultMaxVCPUs
 	}
 
+	if conf.ConfidentialGuest && conf.NumVCPUs != conf.DefaultMaxVCPUs {
+		hvLogger.Warnf("Confidential guests do not support hotplugging of vCPUs. Setting DefaultMaxVCPUs to NumVCPUs (%d)", conf.NumVCPUs)
+		conf.DefaultMaxVCPUs = conf.NumVCPUs
```
Comment (egernst) [inline src/runtime/virtcontainers/hypervisor.go:569]: I'm not sure this is the best policy here -- this isn't really confidential specific -- other VMM configurations also have this limitation.

File: src/runtime/config/configuration-clh.toml.in [hunk 3]
```diff
@@ -15,6 +15,33 @@ path = "@CLHPATH@"
...
 
+# Enable confidential guest support.
+# Toggling that setting may trigger different hardware features, ranging
+# from memory encryption to both memory and CPU-state encryption and integrity.
+# The Kata Containers runtime dynamically detects the available feature set and
+# aims at enabling the largest possible one.
+#
+# Known limitations:
+# * Does not work by design:
+#   - CPU Hotplug
```
Comment (egernst) [inline src/runtime/config/configuration-clh.toml.in:26]: My own curiosity -- is this a policy decision, or is there a limitation in the VMM / guest that prohibits hot plug?

File: src/runtime/virtcontainers/clh.go (hunk 1 above)
Comment (fidencio) [reply src/runtime/virtcontainers/clh.go:660]: Oops, I will raise one PR fixing this, thanks!

File: src/runtime/config/configuration-clh.toml.in (hunk 3 above)
Comment (fidencio) [reply src/runtime/config/configuration-clh.toml.in:26]: My understanding is that it's not supported by the TEE.  It's not a limitation in the VMM / guest, but rather something not supported by Confidential Computing in general.

File: src/runtime/virtcontainers/hypervisor.go (hunk 2 above)
Comment (fidencio) [reply src/runtime/virtcontainers/hypervisor.go:569]: So, from my point either we do this here or we let it error out on the VMM side.
See the answer below, but AFAIU this is a "limitation" of Confidential Computing and not related to the VMM itself.

I am not sure if I follow you on "this isn't confidential specific", mind to expand?

Comment (egernst) [reply src/runtime/virtcontainers/hypervisor.go:569]: Other VMMs and architectures also don't support hot plug (see Firecracker, see baseline aarch64).

If we want to setup a sandbox sizing policy when hot plug isn't available, then we should make it generic -- this isn't a confidential computing specific problem.

Comment (egernst) [reply src/runtime/virtcontainers/hypervisor.go:569]: or, perhaps this is me err; All we're doing is setting default MAX equal to whatever is set for the machine.

Curious -- does CLH and or QEMU fail to boot of max cpu is greater than number of CPU, then?

Comment (fidencio) [reply src/runtime/virtcontainers/hypervisor.go:569]: CLH for sure.  I think QEMU does as well.

Comment (fidencio) [reply src/runtime/virtcontainers/hypervisor.go:569]: /cc @devimc

Comment (devimc) [reply src/runtime/virtcontainers/hypervisor.go:569]: >  does CLH and or QEMU fail to boot of max cpu is greater than number of CPU, then?

yes

File: src/runtime/virtcontainers/clh.go (hunk 1 above)
Comment (fidencio) [reply src/runtime/virtcontainers/clh.go:660]: JFTR: https://github.com/kata-containers/kata-containers/pull/3776

**Pull request merged**